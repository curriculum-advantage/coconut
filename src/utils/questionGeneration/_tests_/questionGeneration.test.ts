import Prando from 'prando';

import {
  itemExists,
  populateArray,
  prandoDigits,
  createPrandoDistributor,
  createPrandoIntGenerator,
  prandoShuffle,
} from '..';

const seed = 'Test_Seed';

describe('itemExists function', () => {
  it('should return true when the given item is included in the given array', () => {
    const exists = itemExists([1], 1);
    expect(exists).toMatchSnapshot();
  });

  it('should return false when the given item is not included in the given array', () => {
    const exists = itemExists([1], 2);
    expect(exists).toMatchSnapshot();
  });
});

// closure to mimic item-generation fn in a predictable manner that can pass unit tests
// because of predictability, length should be <= amount of unique numbers in nonUniqueArray
const createGenerateItemFn = () => {
  let i = -1;
  const nonUniqueArray = [2, 2, 2, 2, 845678, 3, 2, 3, 5, 5, 8, 9, 1, 733, 938];
  return () => {
    i += 1;
    return nonUniqueArray[i];
  };
};

const isDuplicate = (array, item) => array.includes(item);

describe('populateArray function', () => {
  it('should return a non-unique, populated array with the correct length when a isDuplicate callback is not provided', () => {
    const length = 3;
    const generateItem = createGenerateItemFn();
    expect(populateArray(length, generateItem)).toMatchSnapshot();
  });

  it('should return a unique, populated array with the correct length when a isDuplicate callback is provided', () => {
    const length = 6;
    const generateItem = createGenerateItemFn();
    expect(populateArray(length, generateItem, true, isDuplicate)).toMatchSnapshot();
  });
});

describe('createPrandoIntGenerator function', () => {
  const rng = new Prando(seed);
  const getInt = createPrandoIntGenerator(rng, [0, 9]);

  it('should return a callback function', () => {
    expect(typeof getInt).toMatchSnapshot();
  });

  it('should create a callback that returns proportionally distributed integers, and does so repeatedly', () => {
    const distributedIntegers1 = new Array(10)
      .fill(null)
      .reduce(accumulator => [...accumulator, getInt()], []);

    const distributedIntegers2 = new Array(10)
      .fill(null)
      .reduce(accumulator => [...accumulator, getInt()], []);

    expect(distributedIntegers1).toMatchSnapshot();
    expect(distributedIntegers2).toMatchSnapshot();
  });
});

describe('prandoShuffle function', () => {
  it('should return a new shuffled array using items from the given array', () => {
    expect(prandoShuffle(new Prando(seed), [3, 77, 939, 3])).toMatchSnapshot();
  });
});

describe('prandoDigits function', () => {
  const rng = new Prando(seed);

  it('should return an array of deterministically random digits, with the given length', () => {
    const digits = prandoDigits(rng, 8);
    expect(digits.length).toMatchSnapshot();
  });

  it('should return a unique array of deterministically random digits, including zeros, when passed both options', () => {
    const digits = prandoDigits(rng, 10, { unique: true, zeros: true });
    expect(digits).toMatchSnapshot();
  });

  it('should return an array of deterministically random digits with the given range', () => {
    const digits = prandoDigits(rng, 3, { unique: true, start: 1, end: 3 });
    expect(digits).toMatchSnapshot();
  });
});

describe('createPrandoDistributor function', () => {
  const rng = new Prando(seed);
  const types = [[1, 10], [2, 20], [3, 30], [4, 40]];
  const getType = createPrandoDistributor(rng, 10, types);

  it('should return a callback function', () => {
    expect(typeof getType).toMatchSnapshot();
  });

  it('should create a callback that returns a deterministically shuffled, proportionally distributed type', () => {
    const distributedTypes = new Array(10)
      .fill(null)
      .reduce((accumulator, _item, i) => [...accumulator, getType(i)], []);

    expect(distributedTypes).toMatchSnapshot();
  });

  it('should create a callback that returns the same type when passed in a duplicate index', () => {
    const getType2 = createPrandoDistributor(rng, 10, types);
    const type1 = getType2(1);
    const type2 = getType2(1);
    const isSameType = type1 === type2;
    expect(isSameType).toMatchSnapshot();
  });
});
