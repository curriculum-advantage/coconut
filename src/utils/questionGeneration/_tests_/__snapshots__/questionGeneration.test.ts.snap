// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPrandoDistributor function should create a callback that returns a deterministically shuffled, proportionally distributed type 1`] = `
Array [
  3,
  2,
  4,
  2,
  1,
  3,
  3,
  4,
  4,
  4,
]
`;

exports[`createPrandoDistributor function should create a callback that returns the same type when passed in a duplicate index 1`] = `true`;

exports[`createPrandoDistributor function should return a callback function 1`] = `"function"`;

exports[`createPrandoIntGenerator function should create a callback that returns proportionally distributed integers, and does so repeatedly 1`] = `
Array [
  5,
  4,
  8,
  2,
  0,
  3,
  6,
  7,
  9,
  1,
]
`;

exports[`createPrandoIntGenerator function should create a callback that returns proportionally distributed integers, and does so repeatedly 2`] = `
Array [
  0,
  1,
  4,
  8,
  9,
  2,
  5,
  6,
  7,
  3,
]
`;

exports[`createPrandoIntGenerator function should return a callback function 1`] = `"function"`;

exports[`itemExists function should return false when the given item is not included in the given array 1`] = `false`;

exports[`itemExists function should return true when the given item is included in the given array 1`] = `true`;

exports[`populateArray function should return a non-unique, populated array with the correct length when a isDuplicate callback is not provided 1`] = `
Array [
  2,
  2,
  2,
]
`;

exports[`populateArray function should return a unique, populated array with the correct length when a isDuplicate callback is provided 1`] = `
Array [
  2,
  845678,
  3,
  5,
  8,
  9,
]
`;

exports[`prandoDigits function should return a unique array of deterministically random digits, including zeros, when passed both options 1`] = `
Array [
  7,
  3,
  0,
  8,
  1,
  5,
  2,
  6,
  4,
  9,
]
`;

exports[`prandoDigits function should return an array of deterministically random digits with the given range 1`] = `
Array [
  1,
  3,
  2,
]
`;

exports[`prandoDigits function should return an array of deterministically random digits, with the given length 1`] = `8`;

exports[`prandoShuffle function should return a new shuffled array using items from the given array 1`] = `
Array [
  939,
  77,
  3,
  3,
]
`;
